cmd/snap/cmd_run.go

```diff
@@ -702,6 +706,135 @@ func (x *cmdRun) runCmdUnderGdb(origCmd, env []string) error {
+// PID   TIME              SYSCALL
+// 17363 1542815326.700248 execve("/snap/brave/44/usr/bin/update-mime-database", ["update-mime-database", "/home/egon/snap/brave/44/.local/"...], 0x1566008 /* 69 vars */) = 0
+var newExecveRE = regexp.MustCompile(`([0-9]+)\ +([0-9.]+) execve\(\"([^"]+)\"`)
+
+// lines look like:
+// PID   TIME                  SIGNAL
+// 17559 1542815330.242750 --- SIGCHLD {si_signo=SIGCHLD, si_code=CLD_EXITED, si_pid=17643, si_uid=1000, si_status=0, si_utime=0, si_stime=0} ---
+var sigChldTermRE = regexp.MustCompile(`[0-9]+\ +([0-9.]+).*si_pid=([0-9]+),`)
+
+func straceExtractExecRuntime(straceLog string) ([]execRuntime, error) {
+	slog, err := os.Open(straceLog)
+	if err != nil {
+		return nil, err
+	}
+	defer slog.Close()
+
+	execRuntimes := make([]execRuntime, 0, 10)
+	pidToPerf := make(map[string]perfStart)
+
+	r := bufio.NewScanner(slog)
+	for r.Scan() {
+		line := r.Text()
+
+		// look for new Execs
+		match := newExecveRE.FindStringSubmatch(line)
+		if len(match) > 0 {
+			pid := match[1]
+			execStart, err := strconv.ParseFloat(match[2], 64)
+			if err != nil {
+				return nil, err
+			}
+			execve := match[3]
+			// deal with subsequent execve()
+			if perf, ok := pidToPerf[pid]; ok {
+				execRuntimes = append(execRuntimes, execRuntime{
+					Execve:   perf.Execve,
+					TotalSec: execStart - perf.Start,
+				})
+			}
+			pidToPerf[pid] = perfStart{
+				Start:  execStart,
+				Execve: execve,
+			}
+		}
+		match = sigChldTermRE.FindStringSubmatch(line)
+		if len(match) > 0 {
+			sigTime, err := strconv.ParseFloat(match[1], 64)
+			if err != nil {
+				return nil, err
+			}
+			sigPid := match[2]
+			if perf, ok := pidToPerf[sigPid]; ok {
+				execRuntimes = append(execRuntimes, execRuntime{
+					Execve:   perf.Execve,
+					TotalSec: sigTime - perf.Start,
+				})
+				delete(pidToPerf, sigPid)
+			}
+		}
+
+	}
+	if r.Err() != nil {
+		return nil, r.Err()
+	}
+	sort.Sort(byRuntime(execRuntimes))
+
+	return execRuntimes, nil
+}
+
+func displaySortedExecRuntimes(execRuntimes []execRuntime, n int) {
+	if n > len(execRuntimes) {
+		n = len(execRuntimes)
+	}
+
+	fmt.Fprintf(Stderr, "Slowest %d exec calls during snap run:\n", n)
+	for _, rt := range execRuntimes[len(execRuntimes)-n:] {
+		fmt.Fprintf(Stderr, "%2.3f: %s\n", rt.TotalSec, rt.Execve)
+	}
+}
+
+func (x *cmdRun) runCmdWithPerfMonitoring(origCmd, env []string) error {
+	// prepend strace magic
+	cmd, err := straceCmd()
+	if err != nil {
+		return err
+	}
+
+	// FIXME: use e.g. (named?) pipe here and analyze async
+	//straceLog := fmt.Sprintf("/run/snapd/strace-%d.log", os.Getpid())
+	//defer os.Remove(straceLog)
+	straceLog := "/run/snapd/strace.log"
+
+	straceOpts := []string{"-ttt", "-e", "trace=execve", "-o", fmt.Sprintf("%s", straceLog)}
+	cmd = append(cmd, straceOpts...)
+	cmd = append(cmd, origCmd...)
+	// run
+	gcmd := exec.Command(cmd[0], cmd[1:]...)
+	gcmd.Env = env
+	gcmd.Stdin = Stdin
+	gcmd.Stdout = Stdout
+	gcmd.Stderr = Stderr
+	err = gcmd.Run()
```

Comment:
Checking for ineffective assignments
/home/travis/gopath/src/github.com/snapcore/snapd/cmd/snap/cmd_run.go:827:2: ineffectual assignment to err